storage/src/main/java/io/delta/storage/GCSLogStore.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.FileAlreadyExistsException;
+import java.util.Iterator;
+import java.util.concurrent.Callable;
+
+/**
+ * :: Unstable ::
+ * <p>
+ * The [[LogStore]] implementation for GCS, which uses gcs-connector to
+ * provide the necessary atomic and durability guarantees:
+ *
+ * <ol>
+ *   <li>Atomic Visibility: Read/read-after-metadata-update/delete are strongly
+ * consistent for GCS.</li>
+ *
+ *   <li>Consistent Listing: GCS guarantees strong consistency for both object and
+ * bucket listing operations.
+ * https://cloud.google.com/storage/docs/consistency</li>
+ *
+ *   <li>Mutual Exclusion: Preconditions are used to handle race conditions.</li>
+ * </ol>
+ * Regarding file creation, this implementation:
+ * <ul>
+ *    <li>Opens a stream to write to GCS otherwise.</li>
+ *    <li>Throws [[FileAlreadyExistsException]] if file exists and overwrite is false.</li>
+ *    <li>Assumes file writing to be all-or-nothing, irrespective of overwrite option.</li>
+ * </ul>
+ * <p>
+ * This class is not meant for direct access but for configuration based on storage system.
+ * See https://docs.delta.io/latest/delta-storage.html for details.
+ */
+public class GCSLogStore extends HadoopFileSystemLogStore {
+
+    String preconditionFailedExceptionMessage = "412 Precondition Failed";
+
+    public GCSLogStore(Configuration hadoopConf) {
+        super(hadoopConf);
+    }
+
+    @Override
+    public void write(Path path,
+                      Iterator<String> actions,
+                      Boolean overwrite,
+                      Configuration hadoopConf) throws IOException {
+        final FileSystem fs = path.getFileSystem(hadoopConf);
+
+        // This is needed for the tests to throw error with local file system.
+        if (fs instanceof LocalFileSystem && !overwrite && fs.exists(path)) {
+            throw new FileAlreadyExistsException(path.toString());
+        }
+
+        // GCS may upload an incomplete file when the current thread is interrupted, hence we move
+        // the write to a new thread so that the write cannot be interrupted.
+        // TODO Remove this hack when the GCS Hadoop connector fixes the issue.
+        // If overwrite=false and path already exists, gcs-connector will throw
+        // org.apache.hadoop.fs.FileAlreadyExistsException after fs.create is invoked.
+        // This should be mapped to java.nio.file.FileAlreadyExistsException.
+        Callable body = () -> {
+            FSDataOutputStream stream = fs.create(path, overwrite);
+            try {
+                while (actions.hasNext()) {
+                    stream.write((actions.next() + "\n").getBytes(StandardCharsets.UTF_8));
+                }
+            } finally {
+                stream.close();
+            }
+            return "";
+        };
+
+        ThreadUtils.runInNewThread("delta-gcs-logstore-write", true, body);
```

Comment:
where is the try catch for this?

you aren't calling `isPreconditionFailure` anywhere ?

you used to have code here that caught `org.apache.hadoop.fs.FileAlreadyExistsException` and re-threw a `throw new FileAlreadyExistsException(path.toString)` instead?